### ISSUE TITLE
[6] Tent: Fully Test-Time Adaptation by Entropy Minimization

### ISSUE THREAD

**Issue by ejjun92** [ejjun92/Autonomous-Driving-Papers#6]
* Problem: training data (source)와 testing data (target) 간 다를 때 성능 drop -> dataset shift.
이 때, test 하는 동안 새로운 & 다른 데이터 generalize 하기.
즉, adaptation 필요
* Idea: test entropy minimization
  * 즉, prediction에 대한 entropy를 confidence로 잡고 모델 optimize하기.
(entropy는 error와 shift와 연관)
  * entropy를 줄이기 위해,
target data에 대한 inference를 normalize & transform
-> batch by batch로 affine 파라미터 optimize
* Contribution
  * 오직 target 데이터만으로 fully test-time adaptation (no source data)
  * adaptation objective로써 entropy를 사용하여, 모델 예측의 entropy를 줄임으로써 generalization error를 줄일 수 있음을 보임
  * corruption에 대해 강건함을 보임 (ImageNet-C)
  * domain adaptation에서, digit classification과 semantic segmentation에서 online으로 source-free adaptation이 가능함을 보임  
* Evaluation:
  * corruptions for image classification
  * domain shift for digit recognition
  * simulation-to-real shift for semantic segmentation

## Fully Test-time Adaptation
* 기존 adaptation 세팅
  * Transfer learning by fine-tuning
  * Domain adaptation
  * Test-time training
  * Fully test-time adaptation 

* 기존의 방식들과 비교하여, tent의 이점
  * training data와 training loss에 independent. 즉, training 변경 X (domain adaptation, test-time training)
  * target label 필요 X (fine-tuning)

## Method: Test Entropy Minimization via Feature Modulation

* adapt 되는 모델은 (i) supervised task로, (ii) probabilistic & (iii) differentiable하게 train 되어있어야 함 (supervised learning에 대한 일반적인 deep 네트워크 세팅)
  * testing 할 때는 no supervision (supervised task)
  * prediction의 entropy를 측정하려면 prediciton에 대한 distribution이 있어야 함 (probabilistic)
  * fast iterative optimization을 위해 gradient가 요구됨 (differentiable)

### Entropy objective
Shannon entropy 측정.
single prediction을 최적화하는 것은 모든 확률을 가장 가능성 높은 클래스에 할당하는 trivial solution을 가지므로,
이를 방지하기 위해
batch가 공유하는 parameter를 통해 batched prediction에 대해 jointly optimize.

### Modulation parameters
model parameter는 training/source 데이터의 유일한 representation임.
model parameter를 변경하면 모델이 다른 방향으로 발산할 수 있음.
f가 nonlinear하고, model parameter가 고차원이면 test-time 사용에 있어 최적화가 너무 민감하고 비효율적임.

그러므로, 
안전성과 효율성을 위해,
tent에서는 오직 feature modulation만 업데이트함. 
-> 이는 linear (scale & shift)하고, low-dimensional (channel-wise) 함.
아래 2 step modulation
1. statistics에 의한 normalization 
2. parameter에 의한 transformation
구현은 source 모델의 normalization layer 활용.
test 동안에는 normalization statistics와 affine parameters 업데이트.

### 알고리즘
* initialization
  * optimizer가 source model의 각 normalization layer l & channel k에 대해 transformation parameter를 collect
* iteration
  * forward pass: 각 step에서 batch에 대해 normalization statistics & transformation parameter 업데이트. 
  * backward pass: prediction entropy의 gradient에 대해 transformation parameter 업데이트.
    * transformation은 current batch의 prediction 기반이므로, next batch에 영향.
* termination
  * online adaptation: termination 필요 X. test 데이터가 있는 한 iteration 반복.
  * offline adaptation: 모델이 먼저 업데이트 되고, inference 반복. 여러 epoch에 대해 adaptation 계속됨.
